api/app/routers/hfi_calc.py

```diff
@@ -31,8 +34,71 @@
 )
 
 
+def get_prepared_request(
+        session: Session,
+        fire_centre_id: int,
+        start_date: Optional[date],
+        fire_centre_fire_start_ranges: List[FireStartRange]) -> Tuple[HFIResultRequest, bool]:
+    """ Attempt to load the most recent request from the database, failing that creates a new request all
+    set up with default values.
+
+    TODO: request_loaded should become redudant once all requests have been changed.
+    """
+
+    stored_request = get_most_recent_updated_hfi_request(session,
+                                                         fire_centre_id,
+                                                         start_date)
+    request_loaded = False
+    if stored_request:
+        try:
+            result_request = HFIResultRequest.parse_obj(json.loads(stored_request.request))
+            request_loaded = True
+        except ValidationError as validation_error:
+            # This can happen when we change the schema! It's rare - but it happens.
+            logger.error(validation_error)
+    if not request_loaded:
+        # No stored request, so we need to create one.
+        # TODO: selected_station_code_ids make it impossible to have a station selected in one area,
+        # and de-selected in another area. This has to be fixed!
+        selected_station_code_ids = set()
+        planning_area_station_info: Mapping[int, List[StationInfo]] = {}
+        planning_area_fire_starts: Mapping[int, FireStartRange] = {}
+        start_date, end_date = validate_date_range(start_date, None)
+        num_prep_days = (end_date - start_date).days
```

Comment:
won't this lead to an off-by-1 error again? e.g., if a user is requesting a prep period of a single day, end_date = start_date, and `num_prep_days` will equal 0. On the front end we've been adding `+ 1` to the date delta because it's really requesting for the start of the day of `start_date` to the end of the day of `end_date`

Reply:
+1 please write a tests for this

Reply:
Writing a test for this particular edge case is actually a nightmare - there's a function initialize_planning_area_fire_starts, which consumes the number of prep days and ensures the structure is correct. That function is run twice, SOMETIMES - which is just redundant and bad (there's a TODO to that effect).
I hope to re-factor things to address this - right now - reproducing the edge case where this value get's used is too much effort. Instead - I've ripped out the math, and moved it into the schema, so you can do `date_range.num_prep_days()` which ensure the math at least should remain consistent.

Reply:
Nice the date range function should be easy to test then. Which commit was that change made?

Reply:
Just pushed it up now.